client-tests/src/integration.cancellation.test.ts

```diff
@@ -0,0 +1,87 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as assert from 'assert';
+import * as fs from 'fs';
+import * as crypto from 'crypto';
+import * as os from 'os';
+import * as path from 'path';
+
+import * as lsclient from 'vscode-languageclient';
+import * as vscode from 'vscode';
+import { randomBytes } from 'crypto';
+
+suite('Cancellation integration', () => {
+
+	const folderName = randomBytes(21).toString('hex');
+
+	let client!: lsclient.LanguageClient;
+	let middleware: lsclient.Middleware;
+	let uri!: vscode.Uri;
+
+	suiteSetup(async () => {
+		vscode.workspace.registerTextDocumentContentProvider('lsptests', {
+			provideTextDocumentContent: (_uri: vscode.Uri) => {
+				return [
+					'REM @ECHO OFF',
+					'cd c:\source',
+					'REM This is the location of the files that you want to sort',
+					'FOR %%f IN (*.doc *.txt) DO XCOPY c:\source\"%%f" c:\text /m /y',
+					'REM This moves any files with a .doc or',
+					'REM .txt extension from c:\source to c:\textkkk',
+					'REM %%f is a variable',
+					'FOR %%f IN (*.jpg *.png *.bmp) DO XCOPY C:\source\"%%f" c:\images /m /y',
+					'REM This moves any files with a .jpg, .png,',
+					'REM or .bmp extension from c:\source to c:\images;;',
+				].join('\n');
+			}
+		});
+
+		uri = vscode.Uri.parse('lsptests://localhist/test.bat');
+		await vscode.workspace.openTextDocument(uri);
+
+		const serverModule = path.join(__dirname, './servers/testServer.js');
+		const serverOptions: lsclient.ServerOptions = {
+			run: { module: serverModule, transport: lsclient.TransportKind.ipc, cancellationFolderName: folderName },
+			debug: {
+				module: serverModule, transport: lsclient.TransportKind.ipc,
+				cancellationFolderName: folderName, options: { execArgv: ['--nolazy', '--inspect=6014'] }
+			}
+		};
+		const documentSelector: lsclient.DocumentSelector = [{ scheme: 'lsptests', language: 'bat' }];
+
+		middleware = {};
+		const clientOptions: lsclient.LanguageClientOptions = {
+			documentSelector, synchronize: {}, initializationOptions: {}, middleware, cancellationFolderName: folderName
+		};
+
+		client = new lsclient.LanguageClient('css', 'Test Language Server', serverOptions, clientOptions);
+		client.start();
+		await client.onReady();
+
+		assert(fs.existsSync(getFolderForFileBasedCancellation()));
+	});
+
+	suiteTeardown(async () => {
+		await client.stop();
+
+		assert(!fs.existsSync(getFolderForFileBasedCancellation()));
+	});
+
+	test('Confirm Cancellation File', async () => {
+		// TODO: figure out how to test cancellation
```

Comment:
do you have an idea on how I can do this? one thing I can think of is adding TestMethod in testserver and I can call through client.sendRequest directly. would that work?

Reply:
figured out. added integration test